csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -11,130 +11,176 @@
+            Identity identity,
+            Protocol protocol,
+            Current current,
+            CancellationToken cancel) =>
+            _lookupServant.FindObjectByIdAsync(identity, protocol);
+
+        internal Locator(Lookup lookupServant, ILocatorRegistryPrx registry)
         {
-            _lookup = lookup;
+            _lookupServant = lookupServant;
             _registry = registry;
         }
     }
 
     internal class LocatorRegistry : ILocatorRegistry
     {
-        private readonly Dictionary<string, IObjectPrx> _adapters = new ();
+        private readonly Dictionary<(string, Protocol), IObjectPrx> _adapters = new ();
         private readonly object _mutex = new ();
-        private readonly Dictionary<string, HashSet<string>> _replicaGroups = new ();
+        private readonly Dictionary<(string, Protocol), HashSet<string>> _replicaGroups = new ();
 
-        public ValueTask SetAdapterDirectProxyAsync(
+        public void RegisterAdapterEndpoints(
             string adapterId,
-            IObjectPrx? proxy,
+            string replicaGroupId,
+            IObjectPrx endpoints,
             Current current,
             CancellationToken cancel)
         {
             lock (_mutex)
             {
-                if (proxy != null)
+                Protocol protocol = endpoints.Protocol;
+
+                if (replicaGroupId.Length > 0)
                 {
-                    _adapters[adapterId] = proxy.Clone(clearLocator: true, clearRouter: true);
+                    if (!_replicaGroups.TryGetValue((replicaGroupId, protocol), out HashSet<string>? adapterIds))
+                    {
+                        adapterIds = new HashSet<string>();
+                        _replicaGroups.Add((replicaGroupId, endpoints.Protocol), adapterIds);
+                    }
+                    _adapters[(adapterId, protocol)] = endpoints.Clone(clearLocator: true, clearRouter: true);
+                    adapterIds.Add(adapterId);
                 }
                 else
                 {
-                    _adapters.Remove(adapterId);
+                    _adapters[(adapterId, protocol)] = endpoints.Clone(clearLocator: true, clearRouter: true);
                 }
             }
-            return default;
         }
 
-        public ValueTask SetReplicatedAdapterDirectProxyAsync(
+        public void SetAdapterDirectProxy(
+            string adapterId,
+            IObjectPrx? proxy,
+            Current current,
+            CancellationToken cancel)
+        {
+            Debug.Assert(false); // this method is never called since this servant is hosted by an ice2 object adapter.
+            if (proxy != null)
+            {
+                RegisterAdapterEndpoints(adapterId, "", proxy, current, cancel);
+            }
+            else
+            {
+                UnregisterAdapterEndpoints(adapterId, "", Protocol.Ice1, current, cancel);
+            }
```

Comment:
It would be clearer to not provide an implementation that will never be used.

Reply:
Fixed (removed implementation).